### PR TITLE
Feature for sampling and reporting compressibility

### PR DESCRIPTION
**Pull request by shobhitdayal**
This is a feature to sample data-block compressibility and and report them as stats. 1 in N (tunable) blocks is sampled for compressibility using two algorithms:
1. lz4 or snappy for fast compression
2. zstd or zlib for slow but higher compression.

The stats are reported to the caller as raw-bytes and compressed-bytes. The block continues to be compressed for storage using the specified CompressionType.

The db_bench_tool how has a command line option for specifying the sampling rate. It's default value is 0 (no sampling). To test the overhead for a certain value, users can compare the performance of db_bench_tool, varying the sampling rate. It is unlikely to have a noticeable impact for high values like 20.

### CODE REVIEW

File: table/block_based_table_builder.cc [hunk 1]
```diff
@@ -101,83 +101,123 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;  // fall back to no compression.
+      return LZ4_Compress(
+          compression_ctx,
+          GetCompressFormatForVersion(kLZ4Compression, format_version),
+          raw.data(),
+          raw.size(),
+          compressed_output);
```
Comment (siying) [inline table/block_based_table_builder.cc]: Is it done by "make format"? If you haven't done it yet, run it and see.

File: table/block_based_table_builder.cc [hunk 2]
```diff
@@ -101,83 +101,123 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
+  }
+
+  // If requested, we sample one in every N block with a
+  // fast and slow compression algorithm and report the stats.
+  // The users can use these stats to decide if it is worthwhile
+  // enabling compression and they also get a hint about which
+  // compression algorithm wil be beneficial.
+  // FIXME-SD: replace random()
+  if (compression_ctx.sample_for_compression() &&
+      ((random() % compression_ctx.sample_for_compression()) == 0) &&
```
Comment (siying) [inline table/block_based_table_builder.cc]: Though calling % per a block is unlikely to generate noticeable performance regression, but I still feel like it's a good idea to restraint ourselves from using division freely in critical paths. Can you turn it into a multification instead?

File: table/block_based_table_builder.cc [hunk 3]
```diff
@@ -101,83 +101,123 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
+                            c,
+                            format_version,
+                            sampled_output_fast);
+    }
+
+    // Sampling with a slow but high-compression algorithm
+    if (ZSTD_Supported() || Zlib_Supported()) {
+      CompressionType c = ZSTD_Supported() ? kZSTD : kZlibCompression;
+      CompressionContext ctx(c, compression_ctx.sample_for_compression(),
+                             compression_ctx.options());
```
Comment (siying) [inline table/block_based_table_builder.cc]: I'm not sure CompressionContext is where the sample rate should go to. @ajkr will be the best person to determine.

File: table/block_based_table_builder.cc (hunk 1 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: Looking at it.

File: table/block_based_table_builder.cc (hunk 2 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: Will do.

File: table/block_based_table_builder.cc (hunk 3 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: I debated not passing it at all and just choosing a fixed sampling rate internally but I thought it may be better to have an option there. Will talk to @ajkr .

Comment (ajkr) [reply table/block_based_table_builder.cc]: I think putting it inside `CompressionContext` is ok for now. The thing that looks a bit confusing is we have `CompressionContext` with file lifetime (held by `Rep`), and also another `CompressionContext` is created for each call to `CompressBlock`.

After #4849, there'll be a separate structure for per-`CompressBlock` data, i.e., `CompressionInfo`. I can move your sample rate there later.

Comment (ajkr) [reply table/block_based_table_builder.cc]: Well, it looks like `CompressionInfo` will need to be a per-`CompressBlockInternal` struct to work with this PR. Unless it is possible for you to setup the sample rate in the caller to `CompressBlock`? If you want sample rate to only apply to data blocks, `BlockBasedTableBuilder::WriteBlock` might be a good place for it. (Although perhaps I'm just biased towards making merge conflicts easier :)).

File: table/block_based_table_builder.cc [hunk 4]
```diff
@@ -101,83 +102,99 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
-  // Compression method is not supported, or not good compression ratio, so just
-  // fall back to uncompressed form.
+}  // namespace
+
+// format_version is the block format as defined in include/rocksdb/table.h
+Slice CompressBlock(const Slice& raw, const CompressionContext& compression_ctx,
+                    CompressionType* type, uint32_t format_version,
+                    std::string* compressed_output,
+                    std::string* sampled_output_fast,
+                    std::string* sampled_output_slow, bool do_sample) {
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: Move `do_sample` before compressed_output`? Since we put out arguments in the end of argument list.

File: table/table_reader_bench.cc [hunk 5]
```diff
@@ -100,7 +100,7 @@ void TableReaderBenchmark(Options& opts, EnvOptions& env_options,
     tb = opts.table_factory->NewTableBuilder(
         TableBuilderOptions(
             ioptions, moptions, ikc, &int_tbl_prop_collector_factories,
-            CompressionType::kNoCompression, CompressionOptions(),
+            CompressionType::kNoCompression, 0, CompressionOptions(),
```
Comment (riversand963) [inline table/table_reader_bench.cc]: Add a comment for `0`, i.e. `0 /* sample_for_compression */`? This applies to a few other places in this PR as well.

Comment (shobhitdayal) [reply table/table_reader_bench.cc]: Fixed.

File: table/block_based_table_builder.cc (hunk 4 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: do_sample had a default value (false). But it's better the way you suggested. Fixed.

File: table/block_based_table_builder.cc [hunk 6]
```diff
@@ -101,83 +102,100 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
+  // fast and slow compression algorithm and report the stats.
+  // The users can use these stats to decide if it is worthwhile
+  // enabling compression and they also get a hint about which
+  // compression algorithm wil be beneficial.
+  if (do_sample && compression_ctx.sample_for_compression() &&
+      Random::GetTLSInstance()->OneIn(
+          (int)compression_ctx.sample_for_compression()) &&
+      sampled_output_fast && sampled_output_slow) {
+    // Sampling with a fast compression algorithm
+    if (LZ4_Supported() || Snappy_Supported()) {
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: I am a little concerned about hard-coded compression algorithms here. If some one wants to sample other algorithms, does he need to add new code here?

**Comment by riversand963** [comment]
One question. Can you remind me why this sampling needs to exist in the online processing instead of an offline study phase? I think there are RocksDB helper classes to read/decompress SST file blocks. Maybe you can use them to recompress blocks using other algorithms and compare the compression effectiveness. It will be much easier to change the sampling rate without affecting live db.

File: table/block_based_table_builder.cc (hunk 6 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: My previous diff (on internal phabricator) passed in a setting where one could specify the algorithm to use for sampling (just like the algorithm to use for compression). After some internal discussion (see the diff) we decided to sample with one fast and one slow algorithm.

In general, we are looking for an indication for compressibility of data and lz4+snappy on the fast end and Zstd+zlib on the high compression end covers the spectrum pretty well -- new compression algorithms don't really come around that often.

**Comment by shobhitdayal** [comment]
> One question. Can you remind me why this sampling needs to exist in the online processing instead of an offline study phase? I think there are RocksDB helper classes to read/decompress SST file blocks. Maybe you can use them to recompress blocks using other algorithms and compare the compression effectiveness. It will be much easier to change the sampling rate without affecting live db.

Having it online makes it more convenient. The overhead of this is negligible. And even if tool, there will be an overhead, and likely higher if run regularly.

If it is available as a native feature then all users can directly benefit from it. If it is a tool then the tool has to be integrated into the user environment and scripted to run, results aggregated etc. I feel that compression is a core feature in RocksDB and so I think this is a better location for the sampling feature as well.

File: util/compression.h [hunk 7]
```diff
@@ -333,17 +333,24 @@ class CompressionInfo {
...
+        context_(_context),
+        dict_(_dict),
+        type_(_type),
+        sample_for_compression_(_sample_for_compression) {}
 
   const CompressionOptions& options() const { return opts_; }
   const CompressionContext& context() const { return context_; }
   const CompressionDict& dict() const { return dict_; }
   CompressionType type() const { return type_; }
+  uint64_t sampleForCompression() const { return sample_for_compression_; }
```
Comment (riversand963) [inline util/compression.h]: In RocksDB codebase, the naming convention for such function is either `SampleForCompression()` or `sample_for_compression()`.

File: tools/db_bench_tool.cc [hunk 8]
```diff
@@ -2214,7 +2214,7 @@ class Benchmark {
       CompressionOptions opts;
       CompressionContext context(FLAGS_compression_type_e);
       CompressionInfo info(opts, context, CompressionDict::GetEmptyDict(),
-                           FLAGS_compression_type_e);
+                           FLAGS_compression_type_e, 0 /* sample_for_compression */);
```
Comment (riversand963) [inline tools/db_bench_tool.cc]: Can you add a gflags option to specify the value of `sample_for_compression` with 0 as default value? It would be useful to tell us the overhead by running db_bench.

File: db/compaction_job.cc [hunk 9]
```diff
@@ -1498,7 +1498,7 @@ Status CompactionJob::OpenCompactionOutputFile(
       *cfd->ioptions(), *(sub_compact->compaction->mutable_cf_options()),
       cfd->internal_comparator(), cfd->int_tbl_prop_collector_factories(),
       cfd->GetID(), cfd->GetName(), sub_compact->outfile.get(),
-      sub_compact->compaction->output_compression(),
+      sub_compact->compaction->output_compression(), 0,
```
Comment (riversand963) [inline db/compaction_job.cc]: Why set sample rate to 0 here?

**Comment by shobhitdayal** [comment]
> Thanks @shobhitdayal, and can you run `make format`?
> I have taken another pass and left a few comments.
> I also see that you perform sampling during flush (writing to L0), but we care more about compression ration when compacting SST files to the bottommost level, where you hard-code sample rate to be 0. I wonder if it's better to do the other way around, i.e. disable sampling during flush, but enable sampling during bottommost compaction. What's your thought @siying?

Thanks, @riversand963, I'll work through the comments.

In Logdevice there is no level compaction, everything is at L0. But this is a good suggestion for coverage. It might be a good idea to enable sampling during compaction as well.

File: db/compaction_job.cc (hunk 9 above)
Comment (shobhitdayal) [reply db/compaction_job.cc]: This is the compaction path? I assume the compressibility of data is not going to change in the compaction path. Esp. since the data is sampled, there is no bias for what is retained and what is overwritten. I think the compression goodness we'll report can inform  the decision to compress during compaction too.

**Comment by shobhitdayal** [comment]
> > Thanks @shobhitdayal, and can you run `make format`?
> > I have taken another pass and left a few comments.
> > I also see that you perform sampling during flush (writing to L0), but we care more about compression ration when compacting SST files to the bottommost level, where you hard-code sample rate to be 0. I wonder if it's better to do the other way around, i.e. disable sampling during flush, but enable sampling during bottommost compaction. What's your thought @siying?
> 
> Thanks, @riversand963, I'll work through the comments.
> 
> In Logdevice there is no level compaction, everything is at L0. But this is a good suggestion for coverage. It might be a good idea to enable sampling during compaction as well.

Thinking about this some more, the compressibility of data we'll report in the write path should apply to compaction too.

**Comment by shobhitdayal** [comment]
> > > Thanks @shobhitdayal, and can you run `make format`?
> > > I have taken another pass and left a few comments.
> > > I also see that you perform sampling during flush (writing to L0), but we care more about compression ration when compacting SST files to the bottommost level, where you hard-code sample rate to be 0. I wonder if it's better to do the other way around, i.e. disable sampling during flush, but enable sampling during bottommost compaction. What's your thought @siying?
> > 
> > 
> > Thanks, @riversand963, I'll work through the comments.
> > In Logdevice there is no level compaction, everything is at L0. But this is a good suggestion for coverage. It might be a good idea to enable sampling during compaction as well.
> 
> Thinking about this some more, the compressibility of data we'll report in the write path should apply to compaction too.

Make format is reporting nothing. Do I need to collapse the changes or does make format figure it out?
[shobhitdayal@devvm29225.prn1 ~/rocksdb] make format
Makefile:156: Warning: Compiling in debug mode. Don't use the resulting binary in production
build_tools/format-diff.sh
Nothing needs to be reformatted!

**Comment by riversand963** [comment]
> > > > Thanks @shobhitdayal, and can you run `make format`?
> > > > I have taken another pass and left a few comments.
> > > > I also see that you perform sampling during flush (writing to L0), but we care more about compression ration when compacting SST files to the bottommost level, where you hard-code sample rate to be 0. I wonder if it's better to do the other way around, i.e. disable sampling during flush, but enable sampling during bottommost compaction. What's your thought @siying?
> > > 
> > > 
> > > Thanks, @riversand963, I'll work through the comments.
> > > In Logdevice there is no level compaction, everything is at L0. But this is a good suggestion for coverage. It might be a good idea to enable sampling during compaction as well.
> > 
> > 
> > Thinking about this some more, the compressibility of data we'll report in the write path should apply to compaction too.
> 
> Make format is reporting nothing. Do I need to collapse the changes or does make format figure it out?
> [shobhitdayal@devvm29225.prn1 ~/rocksdb] make format
> Makefile:156: Warning: Compiling in debug mode. Don't use the resulting binary in production
> build_tools/format-diff.sh
> Nothing needs to be reformatted!

If `make format` reports nothing, then you do not have to make any explicit formatting change.

**Comment by riversand963** [comment]
> > > Thanks @shobhitdayal, and can you run `make format`?
> > > I have taken another pass and left a few comments.
> > > I also see that you perform sampling during flush (writing to L0), but we care more about compression ration when compacting SST files to the bottommost level, where you hard-code sample rate to be 0. I wonder if it's better to do the other way around, i.e. disable sampling during flush, but enable sampling during bottommost compaction. What's your thought @siying?
> > 
> > 
> > Thanks, @riversand963, I'll work through the comments.
> > In Logdevice there is no level compaction, everything is at L0. But this is a good suggestion for coverage. It might be a good idea to enable sampling during compaction as well.
> 
> Thinking about this some more, the compressibility of data we'll report in the write path should apply to compaction too.

When there are more levels than one, I think the data block written on the write path will be different from the ones written during (bottommost) compaction.

File: tools/db_bench_tool.cc (hunk 8 above)
Comment (shobhitdayal) [reply tools/db_bench_tool.cc]: Fixed.

File: util/compression.h (hunk 7 above)
Comment (shobhitdayal) [reply util/compression.h]: Fixed.

File: db/compaction_job.cc (hunk 9 above)
Comment (riversand963) [reply db/compaction_job.cc]: Discussed offline. Will address this in a separate PR.

Comment (riversand963) [reply db/compaction_job.cc]: Add `/*sample_for_compression*/` after `0`.

File: HISTORY.md [hunk 10]
```diff
@@ -2,6 +2,7 @@
 
 ## 6.0.0 (2/19/2019)
 ### New Features
+* Added data-block sampling for compressibility.
```
Comment (riversand963) [inline HISTORY.md]: Just notice that this should be moved to `Unreleased` section.

Comment (riversand963) [reply HISTORY.md]: Because we have cut the 6.0 release branch already.

File: table/sst_file_writer_collectors.h [hunk 11]
```diff
@@ -33,6 +33,14 @@ class SstFileWriterPropertiesCollector : public IntTblPropCollector {
     return Status::OK();
   }
 
+  virtual void BlockAdd(uint64_t /* blockRawBytes */,
+                        uint64_t /* blockCompressedBytesFast */,
+                        uint64_t /* blockCompressedBytesSlow */) override {
+    // Intentionally left blank. No interest in collecting stats for
+    // blocks.
```
Comment (riversand963) [inline table/sst_file_writer_collectors.h]: `SstFileWriter` is another good place for compression sampling, may be done in a separate PR.

File: tools/db_bench_tool.cc [hunk 12]
```diff
@@ -2183,6 +2185,7 @@ class Benchmark {
 
     auto compression = CompressionTypeToString(FLAGS_compression_type_e);
     fprintf(stdout, "Compression: %s\n", compression.c_str());
+    fprintf(stdout, "Compression sampling rate: %ld\n", FLAGS_sample_for_compression);
```
Comment (riversand963) [inline tools/db_bench_tool.cc]: We use `fprintf(stdout, "Compression sampling rate: %" PRId64 "\n", FLAGS_sample_for_compression)` or `fprintf(stdout, "Compression sampling rate: %" PRIu64 "\n", FLAGS_sample_for_compression)`.

File: tools/db_bench_tool.cc [hunk 13]
```diff
@@ -790,6 +790,8 @@ DEFINE_string(compression_type, "snappy",
 static enum rocksdb::CompressionType FLAGS_compression_type_e =
     rocksdb::kSnappyCompression;
 
+DEFINE_int64(sample_for_compression, 0, "Sample every N block for compression");
```
Comment (riversand963) [inline tools/db_bench_tool.cc]: Since we support setting `sample_for_compression` via command line when running db_bench, could you update the PR description to include an updated test plan with instructions in running db_bench, and show the result of the program execution?

Comment (riversand963) [reply tools/db_bench_tool.cc]: Probably vary the value of `sample_for_compression` to give people an idea of possible overhead.

File: table/block_based_table_builder.cc [hunk 14]
```diff
@@ -43,6 +43,7 @@
 #include "util/compression.h"
 #include "util/crc32c.h"
 #include "util/memory_allocator.h"
+#include "util/random.h"
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: It looks like this header inclusion can be removed.

**Comment by riversand963** [comment]
Thanks @shobhitdayal for the contribution. I think we are almost there.

File: table/block_based_table_builder.cc (hunk 2 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: So I ended up using the interface OneIn(), which was an existing interface and already used int the write path. Turns out it also has a mod. It's highly unlikely that this has any impact given the context we are in. But this might be the right place to make an improvement if needed.

**Comment by shobhitdayal** [comment]
> > > > Thanks @shobhitdayal, and can you run `make format`?
> > > > I have taken another pass and left a few comments.
> > > > I also see that you perform sampling during flush (writing to L0), but we care more about compression ration when compacting SST files to the bottommost level, where you hard-code sample rate to be 0. I wonder if it's better to do the other way around, i.e. disable sampling during flush, but enable sampling during bottommost compaction. What's your thought @siying?
> > > 
> > > 
> > > Thanks, @riversand963, I'll work through the comments.
> > > In Logdevice there is no level compaction, everything is at L0. But this is a good suggestion for coverage. It might be a good idea to enable sampling during compaction as well.
> > 
> > 
> > Thinking about this some more, the compressibility of data we'll report in the write path should apply to compaction too.
> 
> When there are more levels than one, I think the data block written on the write path will be different from the ones written during (bottommost) compaction.

Separate PR, per our discussion. Options:
1. We could mix stats for compressibility in the write path with the compaction path. But that may not be useful given we are interested in the compressibility in the compaction path.
2. We could report compressibility in compaction path as a stat from RocksDB. This also doesn't seem very disirable
3. Update the callback to include level information. Then the caller can decide how to consume and report it as stat. Currently the only caller is Logdevice. But this seems like the best change.

File: table/block_based_table_builder.cc (hunk 14 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: Fixed.

File: tools/db_bench_tool.cc (hunk 12 above)
Comment (shobhitdayal) [reply tools/db_bench_tool.cc]: Fixed.

File: tools/db_bench_tool.cc (hunk 13 above)
Comment (shobhitdayal) [reply tools/db_bench_tool.cc]: Updated the PR.

File: HISTORY.md [hunk 15]
```diff
@@ -1,5 +1,9 @@
 # Rocksdb Change Log
 
+## Unlreleased
```
Comment (riversand963) [inline HISTORY.md]: Typo: Unreleased
Also need to add other sections, i.e. Public API Change, Bug Fixes. Example:
```
#Rocksdb Change Log
## Unreleased
### New Features
<Your feature>
### Public API Change
### Bug fixes
```

Comment (shobhitdayal) [reply HISTORY.md]: Fixed.

File: table/block_based_table_builder.cc [hunk 16]
```diff
@@ -581,10 +617,20 @@ void BlockBasedTableBuilder::WriteBlock(const Slice& raw_block_contents,
...
-    block_contents =
-        CompressBlock(raw_block_contents, compression_info, &type,
-                      r->table_options.format_version, &r->compressed_output);
+                                     *compression_dict, type, sample_for_compression);
+
+    std::string sampled_output_fast;
+    std::string sampled_output_slow;
+    block_contents = CompressBlock(raw_block_contents, compression_info,
+                                   &type, r->table_options.format_version,
+                                   is_data_block, /* do_sample */
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: Nit: `is_data_block /* do_sample */,`. The comma should be between `is_data_block` and the comment.

File: table/block_based_table_builder.cc [hunk 17]
```diff
@@ -101,83 +101,106 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
 
-  // Compression method is not supported, or not good compression ratio, so just
-  // fall back to uncompressed form.
+// format_version is the block format as defined in include/rocksdb/table.h
+Slice CompressBlock(
+    const Slice& raw,
+    const CompressionInfo& info,
+    CompressionType* type,
+    uint32_t format_version,
+    bool do_sample,
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: The `do_sample` argument and `info.sample_for_compression` are redundant.

File: util/compression.h [hunk 18]
```diff
@@ -333,17 +333,24 @@ class CompressionInfo {
   const CompressionContext& context_;
   const CompressionDict& dict_;
   const CompressionType type_;
+  const uint64_t sample_for_compression_;
```
Comment (riversand963) [inline util/compression.h]: Is this necessary?

Comment (shobhitdayal) [reply util/compression.h]: This is the main place where it is stored -- from earlier discussions.

File: table/block_based_table_builder.cc (hunk 17 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: do_sample is just passing in whether it is a data block or not. Sampling is rate. The caller could change the rate before passing in to 0 if they don't want to sample, and then change it back. But this seems cleaner.

File: table/block_based_table_builder.cc (hunk 16 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: Fixed.

File: table/block_based_table_builder.cc (hunk 17 above)
Comment (riversand963) [reply table/block_based_table_builder.cc]: Should we rename it from `do_sample` to `is_data_block`? A sampling rate of `0` indicates sampling is disabled, doesn't it?

File: util/compression.h (hunk 18 above)
Comment (riversand963) [reply util/compression.h]: sg

File: table/block_based_table_builder.cc (hunk 17 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: Alex and I had talked about it. Overriding the sampling rate seemed undesirable. calling it do_sample allows the control to be in the caller. Calling it is_data_block would likely mean a common treatment for all callers. I'll take your recommendation for it.

File: db/compaction_job.cc [hunk 19]
```diff
@@ -1498,7 +1498,7 @@ Status CompactionJob::OpenCompactionOutputFile(
       *cfd->ioptions(), *(sub_compact->compaction->mutable_cf_options()),
       cfd->internal_comparator(), cfd->int_tbl_prop_collector_factories(),
       cfd->GetID(), cfd->GetName(), sub_compact->outfile.get(),
-      sub_compact->compaction->output_compression(),
+      sub_compact->compaction->output_compression(), 0 /*sample_for_compression */,
```
Comment (riversand963) [inline db/compaction_job.cc]: After rethinking about this, I think it's easy to replace the 0 with the following:
```
bottommost_level_ ? sub_compact->compaction->mutable_cf_options()->sample_for_compression : 0
```
Therefore we can sample the compressibility for bottommost level compaction if the user enables it.

File: table/block_based_table_builder.cc [hunk 20]
```diff
@@ -101,83 +101,106 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
+  // compression algorithm wil be beneficial.
+  if (do_sample && info.SampleForCompression() &&
+      Random::GetTLSInstance()->OneIn((int)info.SampleForCompression()) &&
+      sampled_output_fast && sampled_output_slow) {
+    // Sampling with a fast compression algorithm
+    if (LZ4_Supported() || Snappy_Supported()) {
+      CompressionType c = LZ4_Supported() ? kLZ4Compression : kSnappyCompression;
+      CompressionContext context(c);
+      CompressionOptions options;
+      CompressionInfo infoTmp(options, context, CompressionDict::GetEmptyDict(),
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: NIt: `infoTmp` -> `info_tmp`

File: table/block_based_table_builder.cc [hunk 21]
```diff
@@ -101,83 +101,106 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
...
+
+      CompressBlockInternal(raw, infoTmp, format_version, sampled_output_fast);
+    }
+
+    // Sampling with a slow but high-compression algorithm
+    if (ZSTD_Supported() || Zlib_Supported()) {
+      CompressionType c = ZSTD_Supported() ? kZSTD : kZlibCompression;
+      CompressionContext context(c);
+      CompressionOptions options;
+      CompressionInfo infoTmp(options, context, CompressionDict::GetEmptyDict(),
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: ditto

File: options/cf_options.h [hunk 22]
```diff
@@ -243,6 +244,7 @@ struct MutableCFOptions {
   bool paranoid_file_checks;
   bool report_bg_io_stats;
   CompressionType compression;
+  uint64_t sample_for_compression;
```
Comment (riversand963) [inline options/cf_options.h]: Initialize in constructor.

File: db/builder.cc [hunk 23]
```diff
@@ -47,7 +47,7 @@ TableBuilder* NewTableBuilder(
         int_tbl_prop_collector_factories,
     uint32_t column_family_id, const std::string& column_family_name,
     WritableFileWriter* file, const CompressionType compression_type,
-    const CompressionOptions& compression_opts, int level,
+    uint64_t sample_for_compression, const CompressionOptions& compression_opts, int level,
```
Comment (riversand963) [inline db/builder.cc]: I take another look. Seems the argument `sample_for_compression` is redundant, given that `moptions` has a member `sample_for_compression` already.

Comment (riversand963) [reply db/builder.cc]: Never mind, I think we can keep it because we sometimes do not sample for certain files/blocks.

File: db/compaction_job.cc (hunk 19 above)
Comment (riversand963) [reply db/compaction_job.cc]: Let's address this later.

File: db/builder.cc (hunk 23 above)
Comment (riversand963) [reply db/builder.cc]: One question though, what is the intended relationship between `moptions.sample_for_compression` and `sample_for_compression` argument?

File: options/cf_options.h (hunk 22 above)
Comment (shobhitdayal) [reply options/cf_options.h]: Good catch! Fixed.

File: table/block_based_table_builder.cc (hunk 21 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: Fixed.

File: table/block_based_table_builder.cc (hunk 20 above)
Comment (shobhitdayal) [reply table/block_based_table_builder.cc]: Fixed.

File: db/builder.cc (hunk 23 above)
Comment (riversand963) [reply db/builder.cc]: I think we can keep this as is because it might be desirable to have the flexibility of using different sampling rate for different SST files even of the same column family. I do not have a strong opinion on this. :)

**Comment by riversand963** [comment]
@shobhitdayal I have made some changes in two commits. The first commit is merely the result of `make format` on my dev machine. The second commit moves `sample_for_compression` to advanced options where we keep options for advanced use cases. Being able to sample is an advanced use case. PTAL.

**Comment by riversand963** [comment]
Is this PR ready for merge? @shobhitdayal ?

**Comment by shobhitdayal** [comment]
> Is this PR ready for merge? @shobhitdayal ?

Yep, landing now.
